Softswitch, Supervisor and Composer changes

heliosfa opened the pull request:
This PR is a collection of changes that bring the Softswitch, Supervisor and Composer further along.

Headline changes:

- "finish" off Composer so that all of the intentional warnings, etc. are addressed.
- Add file provenance information to the top of all generated source files
- Flesh out the command set to enable more control over the Softswitch build, including a command to allow the operator to specify additional compilation flags
- Add a buffering Softswitch mode so that packets can be formed immediately after RTS executes rather than at the time of sending
- Remove the Supervisor's dependency on PMsg_p
- Remove the pair of uint32_t-packet usage with PMsg_p in Mothership that was preventing packets being sent into devices 
- Fix the typo in Composer introduced in the last merge

Probably a few other things that I have missed as well.

Updated docs attached - these are also on the P drive.
[softswitch Docs.pdf](https://github.com/POETSII/Orchestrator/files/6018662/softswitch.Docs.pdf)
[Packet_Format.pdf](https://github.com/POETSII/Orchestrator/files/6018663/Packet_Format.pdf)

There is a related Docs repository PR: https://github.com/POETSII/orchestrator-documentation/pull/7

mvousden commented on Build/gcc/Resources/orchestrate_template.sh:
```diff
@@ -18,7 +18,7 @@
 export MPICH_PATH="{{ MPICH_DIR }}"
 export PATH="{{ MPICH_DIR }}/bin:{{ RISCV_BIN_DIR }}:$PATH"
-export TRIVIAL_LOG_HANDLER=1
+#export SOFTSWITCH_TRIVIAL_LOG_HANDLER=1
```
I'm confused - I guess we don't want the trivial log handler? According to the box at the bottom of line 42 in the documentation, it's there anyway. Should this be `#export SOFTSWITCH_TRIVIAL_LOG_HANDLER=0`?

heliosfa replied:
The guard is an ifdef, not an if equal. It is only equal to 1  as not setting it 1 did not work in the past.

Essentially, this has moved from the orchestrate template to being configurable by Composer through the Makefile.

mvousden replied:
Removing the commented line looks like the way to go, I think.

heliosfa replied:
👍

mvousden commented on Config/OrchestratorMessages.ocfg:
```diff
@@ -231,10 +231,12 @@
 800(I) : "All composer information cleared."
 801(W) : "Unable to reset a portion of the composer. Did you mean 'clear'?"
-802(W) : "Graph instance '%s' has not been placed. Not %s."
+802(W) : "Graph instance '%s' has not been %s. Not %s."
```
How does this change interact with `Apps_t.cpp:L37`?

heliosfa replied:
Interesting, not sure how I missed that one in my search, but is that POST a mistake anyway? 

Apps_t.cpp:L37 would not have interacted nicely with the old version anyway. I am not sure which message it is meant to be.

mvousden replied:
Scrolling all the way back to `a133b17`, the 802 message was (from the pbuilder section):

```
802(D) : "Removing task %s from the Orchestrator"
```

I think this message is now wrong (we deal with applications, not tasks), and is also missing from both `1.0.0-alpha` and your candidate branch.

I've pushed a change to address this (please check).

heliosfa replied:
> Attached is a review of your documentation. No showstoppers, but still worth a gander, I reckon.
> [0172_gmb_documentation_feedback.txt](https://github.com/POETSII/Orchestrator/files/6049413/0172_gmb_documentation_feedback.txt)

>  - Can you refer to the software address documentation in Section 3.1 of your
>   Packet Format documentation? Or is this document supposed to supercede that?

I thought I had done that with the line `software address as described in the Addresses documentation`, but I can clarify further with a more specific statement?

> Occasionally you use verbatim blocks for file names and variables, and
>  occasionally not. I'd recommend making them all verbatim for consistency,
>   but your call.

Meant to be all of them, but I clearly missed a few. Will try to catch some more.


> Page 4: <Insert punchy intro here> is pretty classic, but maybe something a
>   little more formal would be nice? The current intro feels unstructured and a
>  bit "braindumpy" - consider explaining why you're going along the path you
> have set for the reader.

Honestly, I was fed up with the doc when I got there so didn't write an intro and figured it was better to get the meat out there. I will write an intro.


>  - Page 6: Jam in a reference to Figures 3 and 4 here - better flow.

They are already referenced at the bottom of page 6?a

>  - Page 9: The buffering mode softswitch diagram doesn't appear to call OnSend
>   anywhere. I wonder if there's a good way to visually show the difference
>   between Figures 3 and 4 - it's currently hard to discern a difference. Maybe
>   some red text for bits in Figure 4 that are different from Figure 3?

It doesn't have the OnSend in it - this is shown in the RTS sub in the second "Figure 3" (I have fixed the caption and refs). The only difference is that the OnSend block and the following instrumentation poke are replaced with a memcopy. I have added a ref to Fig 5 in the captions of 3 and 4.

>  - Page 10: "softswitch_onRTS() is called any time there may have been a change
>   in a device's state" is not true - as you point out, OnIdle can modify state
>   and return zero, which won't make onRTS get called.

It could be argued that if state is being changed in OnIdle, it should be returning 1. The e.g. kind of clarifies this.

>  - Page 13 (S2.2.3) there are no handlers in Table II. Do you mean macros?

👍 

> - Page 16 "macroin" (maybe run a spell checker).

I did but probably missed a few as it obviously doesn't like things like "struct", "pinAddr", etc. and I am not adding them all to Word's dictionary!

>  - Page 17: Also holds for Tinsel 0.8.3? Also, is there a delay between when
>  the instrumentation statistics are measured and when the packet is sent?

I believe so, but the clock rate seems to be something that may be changed without us being told...

Instrumentation stats are pulled at the time the packet is sent.

>  - Page 18: Is pktBuf in italics for any particular reason?

I think I was trying to make it stand out slightly as something that only exists in buffering mode.




>  - Page 26: Table XIX has some `a`s that should be `x`s.

Good catch!

> Page 29: Empty parentheses after Composer. 

Not sure what they were doing there.

> Table XXI confuses me - is it showing what headers are included for each translation unit? If so, why are some headers "unassigned"?

It is just meant to be a compact list of source and headers that are needed.




>  - Page 37: I'll be honest, I didn't double check this. I'm happy to take your
   word for it. Also, are the boxes supposed to be connected?

No specific connections, as that won't fit on page.



> More generally, I feel like the pacing of this document could be improved by spending more words on
> synopses to introduce concepts, and by shifting some of the reference-manual-tables into appendices.

Yeah, more synopses could be good. I will have a think!

More to come on the linker stuff in a bit.

heliosfa replied:
👍

mvousden commented on Source/Common/SoftwareAddress.cpp:
```diff
@@ -23,6 +23,12 @@ SoftwareAddress::SoftwareAddress(
 
+/* Alternatively, construct a software address from a uint32_t. */
+SoftwareAddress::SoftwareAddress(uint32_t address){
+    definitions = 10;
```
Why `definitions = 10`? Why not `definitions = SOFTWARE_ADDRESS_FULLY_DEFINED_MASK` (which resolves to 31?) :sweet_potato:

heliosfa replied:
I cannot remember the reason, but there was one at the time.  I am actually struggling to find where I used this new method at the moment. Will investigate further and respond further in due course

mvousden commented on Source/Common/SoftwareAddress.cpp:
```diff
@@ -110,6 +116,12 @@ void SoftwareAddress::set_device(DeviceComponent value)
 
+/* Construct a software address from a uint32_t. */
+void SoftwareAddress::from_uint(uint32_t address){
+    definitions = 10;
```
As above :sweet_potato:

mvousden commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -439,19 +438,31 @@ unsigned Mothership::handle_msg_bend_supr(PMsg_p* message)
+    
+    // Get the input packets.
+    //message->Put<P_Pkt_t>();
+    //message->Get(1, inputPackets);
```
Can you remove these commented lines please, if you've got a moment?

heliosfa replied:
done 👍

mvousden commented on Source/Mothership/MessageUtils.cpp:
```diff
@@ -126,17 +119,15 @@ bool Mothership::decode_packets_message(PMsg_p* message,
-    for (packetIt = packetsBuffer.begin(); packetIt != packetsBuffer.end();
-         packetIt++) packets->push_back(*packetIt);
-
     return true;
```
Most of the changes in this file look like you removing the (asinine) doubled-copies I was doing, so that's good, thank you :+1:.

heliosfa replied:
Yep, that was the intention! part of the tidying

mvousden commented on Source/Mothership/PacketHandlers.cpp:
```diff
@@ -25,7 +25,7 @@ void Mothership::handle_pkt_instr(P_Pkt_t* packet)
 }
 
-/* Handle a packet as an instrumentation packet (see
+/* Handle a packet as a logging packet (see
```
Whoops

mvousden commented on Source/Mothership/ThreadComms.h:
```diff
@@ -128,7 +128,7 @@ class ThreadComms
     std::queue<PMsg_p> MPICncQueue;
     std::queue<PMsg_p> MPIAppQueue;
-    std::queue<std::pair<uint32_t, P_Pkt_t> > BackendOutputQueue;
+    std::queue<P_Addr_Pkt_t> BackendOutputQueue;
```
Good job updating the Mothership documentation accordingly!

mvousden commented on Source/OrchBase/Composer.h:
```diff
@@ -48,6 +49,17 @@ typedef struct devTypStrings_t
+    priInstr
+} ssLoopMode_t;
 
 typedef struct ComposerGraphI_t
```
I know you mention it in the documentation, but a little exposition here about what `ComposerGraphI_t` is for, and how it links to `GraphI_t` via your map, would make it easier to follow what's going on for the uninitated.

heliosfa replied:
Will add something tomorrow or Friday

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -26,14 +27,37 @@ ComposerGraphI_t::ComposerGraphI_t()
     outputDir = "Composer";
```
A bit weird having constructors for a `struct` instead of a `class` in Orchestrator style, but you do you I guess ;p

heliosfa replied:
Yeah, as with most things, it started as a struct, then I needed some behaviour altering stuff.

Structs can have methods and constructors, but I am happy to change it to a class if you prefer.

mvousden replied:
Happy either way

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -26,14 +27,37 @@ ComposerGraphI_t::ComposerGraphI_t()
+    softswitchLoopMode = standard;      // Default to the standard loop mode
+    
+    compilationFlags = "";
+    provenanceCache = "";
```
DRY with other constructor - create a "set defaults" function?

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -58,12 +82,106 @@ void ComposerGraphI_t::clearDevTStrsMap()
+    /* Close breaker and flush the dump. */
+    DumpUtils::close_breaker(file, prefix);
+    fflush(file);
+}
```
Wonderful

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -108,7 +269,254 @@ void Composer::setOutputPath(std::string path)
+int Composer::setBuffMode(GraphI_t* graphI, bool buffMode)
+{
+    ComposerGraphI_t* builderGraphI;
+    //FILE * fd = graphI->par->par->fd;              // Detail output file
```
Not sure of the purpose of this.

heliosfa replied:
The purpose of which bit exactly? the commented out line?

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -108,7 +269,254 @@ void Composer::setOutputPath(std::string path)
+    builderGraphI->compilationFlags += " ";
+    
+    return 0;
 }
```
The above changeset has a lot of code duplication. Perhaps either a master method that splits into smaller methods, or a single multi-keyword-argument method, would be more desirable. I'd give this half a sweet potato if I could, because I care enough to comment about it, but I won't lose any sleep if you change nothing.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -164,11 +572,15 @@ int Composer::generate(GraphI_t* graphI)
     // Get the list of cores
+    if(placer == PNULL)
+    {   // No placer set, we cannot continue.
+        fprintf(fd,"\t***No placer set. Please gently shout at an Orchestrator developer.***\n");
```
I know we shouldn't get to this point, but can you post (even a `001`) instead?

heliosfa replied:
The issue with that is the the Composer doesn't "know" about POST by design. I think we have had a similar discussion before?

mvousden replied:
True, but you can maintain an error stringstream, which is returned to the Handler using the Composer. The handler can then do the posting for you. You could even make it a member of `ComposerGraphI_t` if it's easier to manipulate. This is the approach the Placer takes.

As an aside, the Composer does know about post via `Composer::placer->engine->parent->Post`, but I don't recommend going down that road.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -164,11 +572,15 @@ int Composer::generate(GraphI_t* graphI)
     giToCoresFinder = placer->giToCores.find(graphI);
     if (giToCoresFinder == placer->giToCores.end())
     {   // Something has gone horribly wrong
-        //TODO: Barf
```
:vomiting_face:

heliosfa replied:
Will address

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -360,19 +790,26 @@ int Composer::compile(GraphI_t* graphI)
     ComposerGraphIMap_t::iterator srch = graphIMap.find(graphI);
     if (srch == graphIMap.end())
     {   // The Graph Instance has not been seen before, barf.
+        fprintf(fd,"\tAttempt to compile a non-existant application.\n");
```
Logserver please. (I won't repeat) :sweet_potato:

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -383,12 +820,87 @@ int Composer::compile(GraphI_t* graphI)
+    
+    
+    // Form the Softswitch compilation control string
+    //TODO: Make these const strings at the top
```
Flagging this as a reminder - seems like you've done some of them, but not all.

heliosfa replied:
Yeah, I think this was a "jam tomorrow TODO" for another time, but I can probably address it

mvousden replied:
No rush

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -383,12 +820,87 @@ int Composer::compile(GraphI_t* graphI)
+    
+    switch(builderGraphI->softswitchLogHandler)
+    {   // Control the log handler 
+        case trivial:   makeArgs += "SOFTSWITCH_TRIVIAL_LOG_HANDLER=1 ";
```
How does this interact with the commented change in `orchestrate_template.sh`? :sweet_potato:

heliosfa replied:
This replaces that mechanism.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -383,12 +820,87 @@ int Composer::compile(GraphI_t* graphI)
+        fprintf(fd,"\n");
+        DumpUtils::close_breaker(fd, prefix);
+        fflush(fd);
+        
```
Good man.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -696,13 +1267,13 @@ int Composer::prepareDirectories(ComposerGraphI_t* builderGraphI)
     }
 
-    std::string mkdirSoftswitchPath(taskDir + "/Softswitch");
+    /*std::string mkdirSoftswitchPath(taskDir + "/Softswitch");
```
I'm not sure how this commented code fits in. Is it just old?

heliosfa replied:
Yeah, old code - I have pruned!

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -895,11 +1480,14 @@ int Composer::generateSupervisor(ComposerGraphI_t* builderGraphI)
 
 
-    // Default Properteis, state and message struct content
+    // Default Properteis and state content
```
Typo ("Properteis")

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -914,6 +1502,10 @@ int Composer::generateSupervisor(ComposerGraphI_t* builderGraphI)
+        supervisor_h << "#define BCAST(a)            bcast->a\n";
+        supervisor_h << "#define RTSREPLY()          __rtsReply=true\n";
+        supervisor_h << "#define RTSBCAST()          __rtsBcast=true\n";
         supervisor_h << "\n";
```
A couple of points here - I would've expected these macros to be defined in a separate header file (in the supervisor skeleton) that's included. I'm also not sure why you're appending a `\n`, when a ` ` would result in more readable source. :sweet_potato:

heliosfa replied:
Looking back at my notes, these were put here because they were needed in the generated header, which does not #include any of the other Supervisor headers (there are reasons).

Looking at why it is needed in the generated header, it is because the "code" section is placed in the header. This is placed in the header so that any types from includes in the code may be used in the properties/state. 

This is probably worth having a chat about.

mvousden replied:
Yes, let's have a natter.

heliosfa replied:
Will move to  `SupervisorMacros.h` or equivalent as per voice chat with MLV.

heliosfa replied:
I have ended up moving these to the Supervisor header and addressing the reasons why I did not do this before.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -1048,7 +1653,8 @@ int Composer::generateSupervisor(ComposerGraphI_t* builderGraphI)
+    supervisor_cpp << "int Supervisor::OnImplicit(P_Pkt_t* __inPkt, ";
+    supervisor_cpp << "std::vector<P_Addr_Pkt_t>& __outPkt)\n{\n";
 
     /* TODO remove
```
Flagging this TODO

heliosfa replied:
done

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -1059,25 +1665,71 @@ int Composer::generateSupervisor(ComposerGraphI_t* builderGraphI)
-    supervisor_cpp << "static_cast<const void*>(inMsg->payload));\n";
+    supervisor_cpp << "static_cast<const " << inPktFmt << ">(";
+    supervisor_cpp << "static_cast<const void*>(__inPkt->payload));\n";
     supervisor_cpp << "\tOS_PRAGMA_UNUSED(message)\n\n";
```
As a note, this stuff is really hard to read. I know you know, and I know you know I know there's not enough time, but I thought I'd mention it for posterity. :fireworks:

heliosfa replied:
It is horrible, but short of templating there is not easy way to make it significantly easier to read.

I could split it up onto more lines, but that seems excessive.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -1188,24 +1840,23 @@ void Composer::writeMessageTypes(GraphI_t* graphI, std::ofstream& pkt_h)
 
     pkt_h << "#include <cstdint>\n";
     //pkt_h << "#include \"softswitch_common.h\"\n\n";
```
More comments, which may or may not be useful?

heliosfa replied:
That comment was an old one, the pragmas are there as a reminder to investigate something further.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -1641,7 +2321,7 @@ void Composer::formDevTOutputPinHandlers(devTypStrings_t* dTypStrs)
         {
             handlers_cpp << "   pkt_" << (*pinO)->pMsg->Name();
-            handlers_cpp << "_pyld_t* message = ";
+            handlers_cpp << "_pyld_t* message";
```
Hooray

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -1965,76 +2646,93 @@ void Composer::writeThreadContextInitialiser(P_thread* thread, DevT_t* devT,
+        * If the buffer size is constrained, a warning is generated - if this 
+        * occurs frequently, more graceful handling of rtsBuf overflowing may be 
+        * required.
+        */
```
Well said.

mvousden commented on Source/OrchBase/Composer.cpp:
If not post, you can always write a `#warning` into the generated code.

heliosfa replied:
That is a good point. Added.

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -2373,6 +3072,20 @@ void Composer::writeThreadDevIDefs(ComposerGraphI_t* builderGraphI,
+        devISrch = builderGraphI->devISuperIdxMap.find(devI);
+        if (devISrch == builderGraphI->devISuperIdxMap.end())
+        {   // Something has gone wrong that we are going to ignore.
+            //TODO: Barf
```
Flagging this TODO (is this a good candidate for a `#warning`?)

heliosfa replied:
Yeah, a warning and a fprint is probably in order. Will address

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -2733,7 +3446,7 @@ void Composer::writeDevIInputPinEdgeDefs(GraphI_t* graphI, PinI_t* pinI,
         inEdgePropsI << "_InPin_" << pinI->pT->Name();
-        inEdgePropsI << "_props_t Thread_ ";
+        inEdgePropsI << "_props_t ";
         inEdgePropsI << thrDevName << "_Pin_" << pinI->pT->Name();
```
Hooray

mvousden commented on Source/OrchBase/Composer.h:
```diff
@@ -48,6 +49,17 @@ typedef struct devTypStrings_t
+    standard = 0,
+    priInstr
+} ssLoopMode_t;
 
```
These look a lot like fancy booleans to me, which is fine if you're planning to add more entries.

heliosfa replied:
At the moment, they are. Intention is to expand - we already have a "new" log handler part done, and we want to be able to prioritise sending. I have just not had time to implement them yet, but design is there to support.

mvousden commented on Source/OrchBase/Handlers/CmComp.h:
```diff
@@ -7,6 +7,7 @@
 //==============================================================================
 
 #include <stdio.h>
+#include <algorithm>
```
For `std::transform`, looks like.

heliosfa replied:
Indeed

mvousden commented on Source/OrchBase/Handlers/CmDepl.cpp:
```diff
@@ -41,7 +41,6 @@ void CmDepl::Cm_App(Cli::Cl_t clause)
     /* Ensure that all graph instances are built before proceeding. */
     for (graphIt = graphs.begin(); graphIt != graphs.end(); graphIt++)
     {
-#warning "CmBuil::Cm_Deploy: No logic defined for checking whether an application has been built. Deploy-check does not interact with the Composer."
```
How has this ex-warning been addressed? :sweet_potato:

heliosfa replied:
The GraphInstance has a "built" flag in it that is managed by Composer. CmDepl checks this on (new) line 44

mvousden commented on Source/Softswitch/src/softswitch.cpp:
```diff
@@ -47,7 +49,7 @@ void softswitch_trivial_log_handler(const char* &logStr)
     hdr->swAddr = P_SW_MOTHERSHIP_MASK | P_SW_CNC_MASK;
     hdr->swAddr |= ((P_CNC_LOG << P_SW_OPCODE_SHIFT) & P_SW_OPCODE_MASK);
-    hdr->pinAddr = tinselId(); // usurp Pin Addr for the source HW addr
+    hdr->pinAddr = src;         // The index of the device on the supervisor
```
Good stuff

heliosfa replied:
I have not implemented the needed changes in Mothership yet!

mvousden replied:
That's fine, because they're not currently used (?, he asks tentatively?) :sweet_potato:

heliosfa replied:
They are used for the log handler messages, but the current implementation will just report incorrect thread/core.

heliosfa replied:
Will either modify logserver message or implement needed changes in mothership to avoid user confusion.

heliosfa replied:
> Will ~~either modify logserver message or~~ implement needed changes in mothership to avoid user confusion.

Implemented in e858526b9e02d457eb4b6d27542714d11dd56a96

mvousden commented on Source/Softswitch/src/softswitch_common.cpp:
```diff
@@ -506,12 +514,13 @@ void softswitch_onReceive(ThreadCtxt_t* ThreadContext, volatile void* recv_buf)
     else
     {
-        handler_log(5, "dIDX OOR %d", devAdr);
+        //handler_log(5, "dIDX OOR %d", devAdr);
```
I'm all for keeping comments around if they're useful, but an explanation is helpful.

heliosfa replied:
Addressed in e1b5924dde536e02ebf62243cd62ee007173e730

mvousden commented on Source/Softswitch/src/softswitch_common.cpp:
```diff
@@ -523,48 +532,62 @@ void softswitch_onReceive(ThreadCtxt_t* ThreadContext, volatile void* recv_buf)
-
+            default:    // Unused reserved Opcode - log it.
+                        //handler_log(5, "BAD-OP %d", opcode);
+                        break;
```
But why is this commented out? It looks useful...

heliosfa replied:
Yeah, I need to come back and address these. The "tweaked" handler log requires a `deviceInstance` in context to get the source address. We don't have a 'deviceInstance' in context at this point, so the macro fails. 

I could use the raw __handler_log, with a fabricated source, but we then won't know the source.

I guess I could just pretend to be the first device instance in this case (and the previous one)

mvousden replied:
That's a knotty one - could you reserve a special device ID for this purpose?

heliosfa replied:
That is a possibility, the maximum value?

heliosfa replied:
Will grab the index of the first device on the Softswitch so that we get the correct hardware address. Add comment to source.

 Longer term, a second opcode for softswitch vs. device logging would be good.

heliosfa replied:
> Will grab the index of the first device on the Softswitch so that we get the correct hardware address. Add comment to source.

Addressed in e1b5924dde536e02ebf62243cd62ee007173e730

mvousden commented on Source/Softswitch/src/softswitch_common.cpp:
```diff
@@ -633,16 +656,35 @@ inline uint32_t softswitch_onRTS(ThreadCtxt_t* ThreadContext, devInst_t* device)
+                buffRem = (ThreadContext->rtsEnd >= ThreadContext->rtsStart)   ?
+                        (((ThreadContext->rtsBuffSize - ThreadContext->rtsEnd) +
+                                    ThreadContext->rtsStart))                  :
+                            (ThreadContext->rtsStart - ThreadContext->rtsEnd);
```
I'm not sure, but is this trying to calculate the "amount" (?) of space in the (circular) message buffer? Maybe a comment would help here.

heliosfa replied:
Yes it is, I will add a comment.

mvousden reviewed (changes requested):
A second review - happy with changes so far. Documentation review incoming.

mvousden commented:
Attached is a review of your documentation. No showstoppers, but still worth a gander, I reckon.
[0172_gmb_documentation_feedback.txt](https://github.com/POETSII/Orchestrator/files/6049413/0172_gmb_documentation_feedback.txt)

heliosfa commented:
> Attached is a review of your documentation. No showstoppers, but still worth a gander, I reckon.
> [0172_gmb_documentation_feedback.txt](https://github.com/POETSII/Orchestrator/files/6049413/0172_gmb_documentation_feedback.txt)

>  - Can you refer to the software address documentation in Section 3.1 of your
>   Packet Format documentation? Or is this document supposed to supercede that?

I thought I had done that with the line `software address as described in the Addresses documentation`, but I can clarify further with a more specific statement?

> Occasionally you use verbatim blocks for file names and variables, and
>  occasionally not. I'd recommend making them all verbatim for consistency,
>   but your call.

Meant to be all of them, but I clearly missed a few. Will try to catch some more.


> Page 4: <Insert punchy intro here> is pretty classic, but maybe something a
>   little more formal would be nice? The current intro feels unstructured and a
>  bit "braindumpy" - consider explaining why you're going along the path you
> have set for the reader.

Honestly, I was fed up with the doc when I got there so didn't write an intro and figured it was better to get the meat out there. I will write an intro.


>  - Page 6: Jam in a reference to Figures 3 and 4 here - better flow.

They are already referenced at the bottom of page 6?a

>  - Page 9: The buffering mode softswitch diagram doesn't appear to call OnSend
>   anywhere. I wonder if there's a good way to visually show the difference
>   between Figures 3 and 4 - it's currently hard to discern a difference. Maybe
>   some red text for bits in Figure 4 that are different from Figure 3?

It doesn't have the OnSend in it - this is shown in the RTS sub in the second "Figure 3" (I have fixed the caption and refs). The only difference is that the OnSend block and the following instrumentation poke are replaced with a memcopy. I have added a ref to Fig 5 in the captions of 3 and 4.

>  - Page 10: "softswitch_onRTS() is called any time there may have been a change
>   in a device's state" is not true - as you point out, OnIdle can modify state
>   and return zero, which won't make onRTS get called.

It could be argued that if state is being changed in OnIdle, it should be returning 1. The e.g. kind of clarifies this.

>  - Page 13 (S2.2.3) there are no handlers in Table II. Do you mean macros?

👍 

> - Page 16 "macroin" (maybe run a spell checker).

I did but probably missed a few as it obviously doesn't like things like "struct", "pinAddr", etc. and I am not adding them all to Word's dictionary!

>  - Page 17: Also holds for Tinsel 0.8.3? Also, is there a delay between when
>  the instrumentation statistics are measured and when the packet is sent?

I believe so, but the clock rate seems to be something that may be changed without us being told...

Instrumentation stats are pulled at the time the packet is sent.

>  - Page 18: Is pktBuf in italics for any particular reason?

I think I was trying to make it stand out slightly as something that only exists in buffering mode.




>  - Page 26: Table XIX has some `a`s that should be `x`s.

Good catch!

> Page 29: Empty parentheses after Composer. 

Not sure what they were doing there.

> Table XXI confuses me - is it showing what headers are included for each translation unit? If so, why are some headers "unassigned"?

It is just meant to be a compact list of source and headers that are needed.




>  - Page 37: I'll be honest, I didn't double check this. I'm happy to take your
   word for it. Also, are the boxes supposed to be connected?

No specific connections, as that won't fit on page.



> More generally, I feel like the pacing of this document could be improved by spending more words on
> synopses to introduce concepts, and by shifting some of the reference-manual-tables into appendices.

Yeah, more synopses could be good. I will have a think!

More to come on the linker stuff in a bit.

heliosfa commented:
OK, I think this is closer to a suitable merge. 
Updated docs attached:
* [softswitch Docs.pdf](https://github.com/POETSII/Orchestrator/files/6452582/softswitch.Docs.pdf)
* [softswitch Docs.docx](https://github.com/POETSII/Orchestrator/files/6452584/softswitch.Docs.docx)

heliosfa commented:
fixes #206

heliosfa commented:
fixes #207

heliosfa commented:
fixes #194

mvousden commented on Source/OrchBase/Composer.cpp:
```diff
@@ -812,9 +1458,23 @@ int Composer::generateSupervisor(ComposerGraphI_t* builderGraphI)
+    if(supervisor_bin.fail()) // Check that the file opened
+    {                 // if it didn't, tell logserver and exit
+        //TODO: Barf
+        //par->Post(816, vars_hFName.str(), OSFixes::getSysErrorString(errno));
```
Why not post here? 🍠

heliosfa replied:
This comes back to the "Composer doesn't know about post" and implementing the error string return seemed like a "jam tomorrow" thing given everything else we have on.

mvousden commented on Source/Supervisor/Supervisor.cpp:
```diff
@@ -51,6 +49,32 @@ extern "C"
+        }
+        
+        return Supervisor::DeviceVector[idx].Name;
+    }
```
If I'm understanding correctly, both of these out-of-range outputs (L56 and L67) are checked by the caller, propagating errors appropriately. 🍠 just to make sure.

heliosfa replied:
Yes, that is the intention.

mvousden reviewed (commented):
I think I'm happy. I think there are a couple of outstanding 🍠 s, but they're more questions and quick changes than anything.

🍠: Can you write a quick note explaining how you you've tested your changes?

🍠: Can you try the examples in the `macro-support` branch in the examples repo? Once we've done that, we can open a PR for that branch (I don't want to lose it!)

heliosfa commented:
> Can you write a quick note explaining how you you've tested your changes?

These changes have been tested with the examples that we have, large runs of the heated plate, SIR modelling runs, etc. using XMLs both with and without macros.

Runs have been done both on the hardware and MPI backend, with transitions observed with a debugger.


> Can you try the examples in the `macro-support` branch in the examples repo? Once we've done that, we can open a PR for that branch (I don't want to lose it!)

The ones I have tried (a selection) work as expected.

mvousden reviewed (approved):
My :sweet_potato:s have been addressed.

heliosfa commented:
Merged and the docs on P have been updated.

Once we have this spate of PRs in, I will update the P drive version of the Orchestrator
